internal/framework/status/gateway.go

```diff
@@ -1,16 +1,115 @@
 package status
 
 import (
+	"context"
+	"fmt"
 	"sort"
 
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/nginxinc/nginx-gateway-fabric/internal/framework/helpers"
+	"github.com/nginxinc/nginx-gateway-fabric/internal/mode/static/config"
 )
 
+// GetGatewayAddresses gets the addresses for the Gateway.
+func GetGatewayAddresses(
+	ctx context.Context,
+	k8sClient client.Client,
+	svc *v1.Service,
+	podConfig config.GatewayPodConfig,
+) ([]v1beta1.GatewayStatusAddress, error) {
+	podAddress := []v1beta1.GatewayStatusAddress{
+		{
+			Type:  helpers.GetPointer(v1beta1.IPAddressType),
+			Value: podConfig.PodIP,
+		},
+	}
+
+	var gwSvc v1.Service
+	if svc == nil {
+		key := types.NamespacedName{Name: podConfig.ServiceName, Namespace: podConfig.Namespace}
+		if err := k8sClient.Get(ctx, key, &gwSvc); err != nil {
+			return podAddress, fmt.Errorf("error finding Service for Gateway: %w", err)
+		}
+	} else {
+		gwSvc = *svc
+	}
+
+	var addresses, hostnames []string
+	switch gwSvc.Spec.Type {
+	case v1.ServiceTypeNodePort:
+		var err error
+		addresses, err = getNodeAddresses(ctx, k8sClient)
+		if err != nil {
+			return podAddress, fmt.Errorf("error getting Node addresses: %w", err)
+		}
+	case v1.ServiceTypeLoadBalancer:
+		for _, ingress := range gwSvc.Status.LoadBalancer.Ingress {
+			if ingress.IP != "" {
+				addresses = append(addresses, ingress.IP)
+			} else if ingress.Hostname != "" {
+				hostnames = append(hostnames, ingress.Hostname)
+			}
+		}
+	}
+
+	gwAddresses := make([]v1beta1.GatewayStatusAddress, 0, len(addresses)+len(hostnames))
+	for _, addr := range addresses {
+		statusAddr := v1beta1.GatewayStatusAddress{
+			Type:  helpers.GetPointer(v1beta1.IPAddressType),
+			Value: addr,
+		}
+		gwAddresses = append(gwAddresses, statusAddr)
+	}
+
+	for _, hostname := range hostnames {
+		statusAddr := v1beta1.GatewayStatusAddress{
+			Type:  helpers.GetPointer(v1beta1.HostnameAddressType),
+			Value: hostname,
+		}
+		gwAddresses = append(gwAddresses, statusAddr)
+	}
+
+	return gwAddresses, nil
+}
+
+func getNodeAddresses(
+	ctx context.Context,
+	k8sClient client.Client,
+) ([]string, error) {
+	var nodeList v1.NodeList
+	if err := k8sClient.List(ctx, &nodeList); err != nil {
+		return nil, err
+	}
+
+	nodeAddresses := make([]string, 0, len(nodeList.Items))
+	for _, node := range nodeList.Items {
+		var externalIP, internalIP string
+		for _, address := range node.Status.Addresses {
```

Comment:
it gets a bit complicated if the service will have https://github.com/nginxinc/nginx-gateway-fabric/blob/5324908e6e1145bec5f2f0ab80b312a809ad1744/deploy/manifests/service/loadbalancer.yaml#L13 , in that case  NGF will only be available through nodes where a NGF pod is running, not all nodes